### PR TITLE
WCAG 2.1 Development

### PR DESCRIPTION
**Pull request by jestonihpi**
## Description

Complete initial accessibility audit on codebase.

Resolves https://github.com/GravityPDF/gravity-pdf/issues/1028

## Testing instructions
These changes touch many places in the codebase, so checking there is no display issues or functionality problems will be important (the E2E tests should catch these though, if they were all working...)

## Checklist:
- [X] I've tested the code.
- [X] My code is easy to read, follow, and understand
- [X] My code follows the accessibility standards. 
- [X] My code has proper inline documentation / docblocks.

### CODE REVIEW

**Comment by jakejackson1** [comment]
@jestonihpi the big standout to me right now is none of the strings you've added are translatable. Have a ready of https://developer.wordpress.org/plugins/internationalization/how-to-internationalize-your-plugin/ to find out more about how this is handled in WordPress. 

For the JS, you would pass the strings to our localize function here: https://github.com/GravityPDF/gravity-pdf/blob/development/src/Helper/Helper_Data.php#L215-L319

File: src/Helper/Helper_PDF_List_Table.php [hunk 1]
```diff
@@ -332,7 +336,7 @@ public function column_name( $item ) {
 
 		?>
 
-		<a href="<?= $edit_url; ?>"><strong><?= rgar( $item, 'name' ); ?></strong></a>
+		<a href="<?= $edit_url; ?>" aria-label="<?= $pdf_name ?> PDF"><strong><?= $pdf_name ?></strong></a>
```
Comment (jakejackson1) [inline src/Helper/Helper_PDF_List_Table.php:339]: Make translatable. Also need to run attributes through `esc_attr()` for security https://developer.wordpress.org/reference/functions/esc_attr/

File: src/Helper/Helper_PDF_List_Table.php [hunk 2]
```diff
@@ -361,6 +365,6 @@ public function column_name( $item ) {
 	 * @since 4.0
 	 */
 	public function no_items() {
-		printf( esc_html__( "This form doesn't have any PDFs. Let's go %1\$screate one%2\$s.", 'gravity-forms-pdf-extended' ), "<a href='" . add_query_arg( [ 'pid' => 0 ] ) . "'>", '</a>' );
+		printf( esc_html__( "%3\$sThis form doesn't have any PDFs. Let's go %1\$screate one%2\$s.%4\$s", 'gravity-forms-pdf-extended' ), "<a href='" . add_query_arg( [ 'pid' => 0 ] ) . "'>", '</a>', '<label>', '</label>' );
```
Comment (jakejackson1) [inline src/Helper/Helper_PDF_List_Table.php:368]: Don't want to complicate the translation string further with more `printf` placeholders Instead, wrap the entire output with your `<span>` tag.

File: src/View/html/FormSettings/list.php [hunk 3]
```diff
@@ -34,7 +34,7 @@
 				<div class="tablenav top">
 					<div class="alignleft actions bulkactions"></div>
 					<div class="alignright">
-						<a class="button" href="<?= esc_url( $args['add_new_url'] ) ?>"><?php esc_html_e( 'Add New', 'gravityforms' ); ?></a>
+						<a class="button" href="<?= esc_url( $args['add_new_url'] ) ?>" aria-label="Add new PDF"><?php esc_html_e( 'Add New', 'gravityforms' ); ?></a>
```
Comment (jakejackson1) [inline src/View/html/FormSettings/list.php:37]: Needs to be translatable

File: src/View/html/GravityForms/fieldset.php [hunk 4]
```diff
@@ -44,7 +44,8 @@
 			</legend>
 
 			<span class="gform-settings-panel__collapsible-control">
-				<input type="checkbox" class="gform-settings-panel__collapsible-toggle-checkbox" name="<?= esc_attr( $collapsible_name ) ?>" id="<?= esc_attr( $collapsible_name ) ?>" value="1" onclick="this.checked ? this.closest( '.gform-settings-panel' ).classList.add( 'gform-settings-panel--collapsed' ) : this.closest( '.gform-settings-panel' ).classList.remove( 'gform-settings-panel--collapsed' )" <?php checked( empty( $args['collapsible-open'] ) ); ?>>
+				<input type="checkbox" class="gform-settings-panel__collapsible-toggle-checkbox" name="<?= esc_attr( $collapsible_name ) ?>" id="<?= esc_attr( $collapsible_name ) ?>" value="1" onclick="this.checked ? this.closest( '.gform-settings-panel' ).classList.add( 'gform-settings-panel--collapsed' ) : this.closest( '.gform-settings-panel' ).classList.remove( 'gform-settings-panel--collapsed' )" onkeydown="if(event.keyCode==13){ event.preventDefault(); this.click(); }"
+					<?php checked( empty( $args['collapsible-open'] ) ); ?>>
```
Comment (jakejackson1) [inline src/View/html/GravityForms/fieldset.php:48]: Best to have each attribute begin on a new line for easier scanning.

File: src/View/html/PDF/entry_detailed_pdf.php [hunk 5]
```diff
@@ -19,30 +19,30 @@
 ?>
 
 <?php do_action( 'gfpdf_entry_detail_pre_markup', $args['pdfs'] ); ?>
+<ul>
+	<?php foreach ( $args['pdfs'] as $pdf ): ?>
+		<?php do_action( 'gfpdf_entry_detail_pre_container_markup', $pdf, $args['pdfs'] ); ?>
 
-<?php foreach ( $args['pdfs'] as $pdf ): ?>
-	<?php do_action( 'gfpdf_entry_detail_pre_container_markup', $pdf, $args['pdfs'] ); ?>
+		<li class="gfpdf_detailed_pdf_container" >
```
Comment (jakejackson1) [inline src/View/html/PDF/entry_detailed_pdf.php:26]: Remove space before closing `>`

File: src/View/html/PDF/entry_detailed_pdf.php [hunk 6]
```diff
@@ -19,30 +19,30 @@
...
+			<?php do_action( 'gfpdf_entry_detail_pre_pdf_name_markup', $pdf, $args['pdfs'] ); ?>
 
-		<?php do_action( 'gfpdf_entry_detail_pre_pdf_name_markup', $pdf, $args['pdfs'] ); ?>
+			<div><?= esc_html( $pdf['name'] ); ?></div>
 
-		<div><?= esc_html( $pdf['name'] ); ?></div>
+			<?php do_action( 'gfpdf_entry_detail_post_pdf_name_markup', $pdf, $args['pdfs'] ); ?>
 
-		<?php do_action( 'gfpdf_entry_detail_post_pdf_name_markup', $pdf, $args['pdfs'] ); ?>
+			<div class="gfpdf_detailed_pdf_cta" aria-label="View or Download <?= $pdf['settings']['filename'] ?>.pdf ">
```
Comment (jakejackson1) [inline src/View/html/PDF/entry_detailed_pdf.php:34]: make translatable

File: src/assets/js/react/components/CoreFonts/CoreFontListResults.js [hunk 7]
```diff
@@ -96,8 +97,11 @@ export class Retry extends React.Component {
         data-test='component-retry-link'
         href='#'
         onClick={this.handleTriggerRetryFontDownload}
+        aria-live='polite'
+        role='dialog'
       >
         {this.props.retryText}
+
```
Comment (jakejackson1) [inline src/assets/js/react/components/CoreFonts/CoreFontListResults.js:104]: revert

File: src/assets/js/react/components/FontManager/AdvancedButton.js [hunk 8]
```diff
@@ -50,7 +50,7 @@ export class AdvancedButton extends React.Component {
         className='button gfpdf-button'
         onClick={this.handleClick}
       >
-        Advanced
+        Manage
```
Comment (jakejackson1) [inline src/assets/js/react/components/FontManager/AdvancedButton.js:53]: make translatable

File: src/assets/js/react/components/FontManager/FontListItems.js [hunk 9]
```diff
@@ -329,7 +329,7 @@ export class FontListItems extends Component {
     const list = !searchResult ? fontList : searchResult
 
     return (
-      <div data-test='component-FontListItems' className='font-list-items'>
+      <div data-test='component-FontListItems' className='font-list-items' role='listbox' aria-label='Installed fonts.' aria-live="polite">
```
Comment (jakejackson1) [inline src/assets/js/react/components/FontManager/FontListItems.js:332]: make translatable

File: src/assets/js/react/components/Template/TemplateActivateButton.js [hunk 10]
```diff
@@ -56,6 +56,7 @@ export class TemplateActivateButton extends React.Component {
         href='#'
         tabIndex='150'
         className='button activate'
+        aria-label={ this.props.buttonText + ' template' }
```
Comment (jakejackson1) [inline src/assets/js/react/components/Template/TemplateActivateButton.js:59]: make translatable

File: src/assets/js/react/components/Template/TemplateButton.js [hunk 11]
```diff
@@ -50,8 +50,9 @@ class TemplateButton extends React.Component {
         className='button gfpdf-button'
         onClick={this.handleClick}
         ref={node => (this.button = node)}
+        aria-label='Manage PDF Templates'
       >
-        {this.props.buttonText}
+       Manage
```
Comment (jakejackson1) [inline src/assets/js/react/components/Template/TemplateButton.js:55]: make translatable

File: src/assets/js/react/components/Template/TemplateButton.js [hunk 12]
```diff
@@ -50,8 +50,9 @@ class TemplateButton extends React.Component {
         className='button gfpdf-button'
         onClick={this.handleClick}
         ref={node => (this.button = node)}
+        aria-label='Manage PDF Templates'
```
Comment (jakejackson1) [inline src/assets/js/react/components/Template/TemplateButton.js:53]: make translatable

File: src/assets/js/react/components/Template/TemplateDeleteButton.js [hunk 13]
```diff
@@ -104,6 +104,7 @@ export class TemplateDeleteButton extends React.Component {
         href='#'
         tabIndex='150'
         className='button button-secondary delete-theme ed_button'
+        aria-label={this.props.buttonText + ' template'}
```
Comment (jakejackson1) [inline src/assets/js/react/components/Template/TemplateDeleteButton.js:107]: make translatable

File: src/assets/js/react/components/Template/TemplateListItem.js [hunk 14]
```diff
@@ -86,6 +86,8 @@ export class TemplateListItem extends React.Component {
         className={activeTemplate}
         data-slug={item.id}
         tabIndex='150'
+        role="option"
+        aria-label={item.group + ' ' + item.template + ' details' }
```
Comment (jakejackson1) [inline src/assets/js/react/components/Template/TemplateListItem.js:90]: make translatable

**Review by jakejackson1 — CHANGES_REQUESTED** [review]
Mostly just translation-related issues that need to be addressed.

**Comment by jestonihpi** [comment]
> @jestonihpi the big standout to me right now is none of the strings you've added are translatable. Have a ready of https://developer.wordpress.org/plugins/internationalization/how-to-internationalize-your-plugin/ to find out more about how this is handled in WordPress.
> 
> For the JS, you would pass the strings to our localize function here first so you could then access it in JS using the GFPDF object: https://github.com/GravityPDF/gravity-pdf/blob/development/src/Helper/Helper_Data.php#L215-L319

Copy Jake. I will add them right away.

File: src/Helper/Helper_Options_Fields.php [hunk 15]
```diff
@@ -237,7 +236,8 @@ public function get_registered_fields() {
 				'uninstaller' => [
 					'id'         => 'uninstaller',
 					'name'       => esc_html__( 'Uninstall Gravity PDF', 'gravity-forms-pdf-extended' ),
-					'desc'       => '<p class="alert error">' . esc_html__( 'This operation deletes ALL Gravity PDF settings and deactivates the plugin. If you continue, all settings, configuration, custom templates and fonts will be removed.', 'gravity-forms-pdf-extended' ) . '</p>',
+					'desc'       => '<p class="alert error" id="uninstall-gfpdf-description">' . esc_html__( 'This operation deletes ALL Gravity PDF settings and deactivates the plugin. If you continue, all settings, configuration, custom templates and fonts will be removed.', 'gravity-forms-pdf-extended' ) . '</p>',
+
```
[the remark this answers is not in the document]
Comment (jestonihpi) [reply src/Helper/Helper_Options_Fields.php:240]: removed

File: src/Helper/Helper_Options_Fields.php [hunk 16]
```diff
@@ -77,7 +77,6 @@ public function get_registered_fields() {
 						'type'    => 'select',
 						'options' => $this->get_installed_fonts(),
 						'tooltip' => '<h6>' . esc_html__( 'Fonts', 'gravity-forms-pdf-extended' ) . '</h6>' . esc_html__( 'Gravity PDF comes bundled with fonts for most languages world-wide. Want to use a specific font type? Use the font installer (found in the Tools tab).', 'gravity-forms-pdf-extended' ),
-						'class'   => 'gfpdf-font-manager',
```
[the remark this answers is not in the document]
Comment (jestonihpi) [reply src/Helper/Helper_Options_Fields.php:80]: Updated.

File: src/Helper/Helper_Abstract_Options.php [hunk 17]
```diff
@@ -1656,6 +1656,7 @@ public function license_callback( $args ) {
 			$html .= sprintf( $alert, wp_kses_post( $value['msg'] ) );
 		}
 
+		$html .= '<label for="gfpdf_settings[' . $args['id'] . ']" class="screen-reader-text"> ' . $args['name'] . '  license key</label>';
```
[the remark this answers is not in the document]
Comment (jestonihpi) [reply src/Helper/Helper_Abstract_Options.php:1659]: Updated

File: src/Helper/Helper_Options_Fields.php [hunk 18]
```diff
@@ -325,14 +325,13 @@ public function get_registered_fields() {
...
 
 					'custom_pdf_size' => [
-						'id'       => 'custom_pdf_size',
-						'name'     => esc_html__( 'Custom Paper Size', 'gravity-forms-pdf-extended' ),
-						'desc'     => esc_html__( 'Control the exact paper size. Can be set in millimeters or inches.', 'gravity-forms-pdf-extended' ),
-						'type'     => 'paper_size',
-						'size'     => 'small',
-						'class'    => 'gfpdf-hidden gfpdf_paper_size_other',
-						'std'      => $this->get_option( 'default_custom_pdf_size' ),
-						'required' => true,
```
[the remark this answers is not in the document]
Comment (jestonihpi) [reply src/Helper/Helper_Options_Fields.php:335]: Reverted

File: src/Helper/Helper_PDF_List_Table.php [hunk 19]
```diff
@@ -135,7 +135,7 @@ public function display() {
 		$singular = rgar( $this->_args, 'singular' );
 		?>
 
-		<table class="wp-list-table <?= implode( ' ', $this->get_table_classes() ); ?>" cellspacing="0">
+		<table class="wp-list-table <?= implode( ' ', $this->get_table_classes() ); ?>" cellspacing="0" aria-label="PDF List">
```
[the remark this answers is not in the document]
Comment (jestonihpi) [reply src/Helper/Helper_PDF_List_Table.php:138]: Updated

File: src/Helper/Helper_PDF_List_Table.php [hunk 20]
```diff
@@ -232,7 +232,6 @@ public function column_cb( $item ) {
 			<svg viewBox="0 0 6 6" xmlns="http://www.w3.org/2000/svg"><circle cx="3" cy="2" r="1" stroke-width="2"/></svg>
 			<span class="gform-status-indicator-status"><?php echo esc_html( $text ); ?></span>
 		</button>
-
```
[the remark this answers is not in the document]
Comment (jestonihpi) [reply src/Helper/Helper_PDF_List_Table.php:235]: removed

File: src/Helper/Helper_PDF_List_Table.php [hunk 21]
```diff
@@ -246,7 +245,7 @@ public function column_cb( $item ) {
 	 */
 	public function column_notifications( $item ) {
 		if ( ! isset( $item['notification'] ) || count( $item['notification'] ) === 0 ) {
-			esc_html_e( 'None', 'gravity-forms-pdf-extended' );
+			printf( '<span>%s</span>', esc_html( 'None', 'gravity-forms-pdf-extended' ) );
```
[the remark this answers is not in the document]
Comment (jestonihpi) [reply src/Helper/Helper_PDF_List_Table.php:248]: Updated

File: src/Helper/Helper_PDF_List_Table.php [hunk 22]
```diff
@@ -276,13 +276,17 @@ public function column_shortcode( $item ) {
...
 		 */
-		$name = str_replace( '"', '', $item['name'] );
+		$name   = str_replace( '"', '', $item['name'] );
+		$pdf_id = $item['id'];
 
 		/* Prepare our shortcode sample */
 		$shortcode = '[gravitypdf name="' . esc_attr( $name ) . '" id="' . esc_attr( $item['id'] ) . '" text="' . esc_attr__( 'Download PDF', 'gravity-forms-pdf-extended' ) . '"]';
 
+		/* Set up hidden text for screen readers */
+		echo '<p id="' . $pdf_id . '" class="screen-reader-text" >Copy Download ' . $item['name'] . ' Shortcode</p>';
```
[the remark this answers is not in the document]
Comment (jestonihpi) [reply src/Helper/Helper_PDF_List_Table.php:286]: Updated

File: src/Helper/Helper_PDF_List_Table.php [hunk 23]
```diff
@@ -302,9 +306,8 @@ public function column_template( $item ) {
 			$template_group = $template['group'];
 			$template_name  = $this->templates->maybe_add_template_compatibility_notice( $template['template'], $template['required_pdf_version'] );
 
-			?>
-			<strong><?= $template_group; ?></strong> <?= $template_name; ?>
-			<?php
+			echo sprintf( '<span aria-label="%1$s %2$s"><strong>%1$s</strong> %2$s</span>', $template_group, $template_name );
```
[the remark this answers is not in the document]
Comment (jestonihpi) [reply src/Helper/Helper_PDF_List_Table.php:309]: Updated.

File: src/Helper/Helper_PDF_List_Table.php [hunk 24]
```diff
@@ -276,13 +276,17 @@ public function column_shortcode( $item ) {
 		 * While esc_attr() used below will ensure no display issues when copied the double quote will cause shortcode parse issues
 		 * We'll prevent this by removing them before hand
 		 */
-		$name = str_replace( '"', '', $item['name'] );
+		$name   = str_replace( '"', '', $item['name'] );
+		$pdf_id = $item['id'];
```
[the remark this answers is not in the document]
Comment (jestonihpi) [reply src/Helper/Helper_PDF_List_Table.php:280]: Updated.

File: src/Helper/Helper_Options_Fields.php [hunk 25]
```diff
@@ -353,7 +352,7 @@ public function get_registered_fields() {
 						'options' => $this->get_installed_fonts(),
 						'std'     => $this->get_option( 'default_font' ),
 						'desc'    => sprintf( esc_html__( 'Set the primary font used in PDFs. You can also %1$sinstall your own%2$s.', 'gravity-forms-pdf-extended' ), '<a href="' . $this->data->settings_url . '&tab=tools#manage_fonts">', '</a>' ),
-						'class'   => 'gfpdf_font_type gfpdf-font-manager',
```
[the remark this answers is not in the document]
Comment (jestonihpi) [reply src/Helper/Helper_Options_Fields.php:356]: Updated

File: src/Helper/Helper_PDF_List_Table.php (hunk 1 above)
Comment (jestonihpi) [reply src/Helper/Helper_PDF_List_Table.php:339]: Updated.

File: src/Helper/Helper_PDF_List_Table.php (hunk 2 above)
Comment (jestonihpi) [reply src/Helper/Helper_PDF_List_Table.php:368]: Updated.

File: src/View/html/FormSettings/list.php (hunk 3 above)
Comment (jestonihpi) [reply src/View/html/FormSettings/list.php:37]: Updated.

File: src/View/html/GravityForms/fieldset.php (hunk 4 above)
Comment (jestonihpi) [reply src/View/html/GravityForms/fieldset.php:48]: Added new line on both enter and click events.

File: src/View/html/PDF/entry_detailed_pdf.php (hunk 5 above)
Comment (jestonihpi) [reply src/View/html/PDF/entry_detailed_pdf.php:26]: Updated

File: src/View/html/PDF/entry_detailed_pdf.php (hunk 6 above)
Comment (jestonihpi) [reply src/View/html/PDF/entry_detailed_pdf.php:34]: Updated.

File: src/assets/js/react/components/FontManager/AdvancedButton.js (hunk 8 above)
Comment (jestonihpi) [reply src/assets/js/react/components/FontManager/AdvancedButton.js:53]: Updated.

File: src/assets/js/react/components/FontManager/FontListItems.js (hunk 9 above)
Comment (jestonihpi) [reply src/assets/js/react/components/FontManager/FontListItems.js:332]: Updated.

File: src/assets/js/react/components/Template/TemplateActivateButton.js (hunk 10 above)
Comment (jestonihpi) [reply src/assets/js/react/components/Template/TemplateActivateButton.js:59]: Updated.

File: src/assets/js/react/components/Template/TemplateButton.js (hunk 11 above)
Comment (jestonihpi) [reply src/assets/js/react/components/Template/TemplateButton.js:55]: Updated.

File: src/assets/js/react/components/Template/TemplateButton.js (hunk 12 above)
Comment (jestonihpi) [reply src/assets/js/react/components/Template/TemplateButton.js:53]: Updated.

File: src/assets/js/react/components/Template/TemplateDeleteButton.js (hunk 13 above)
Comment (jestonihpi) [reply src/assets/js/react/components/Template/TemplateDeleteButton.js:107]: Updated.

File: src/assets/js/react/components/Template/TemplateListItem.js (hunk 14 above)
Comment (jestonihpi) [reply src/assets/js/react/components/Template/TemplateListItem.js:90]: Updated.

**Review by jestonihpi — COMMENTED** [review]
Pushed an updates for this review @jakejackson1

**Comment by jakejackson1** [comment]
@debbieellis7 I've got this PR in working order, and I'm happy with the accessibility changes. Can you check it over and provide any comments or concerns you might have before merging? Thanks!

**Pull request merged**